pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -0,0 +1,215 @@
+"""Plugin for constructor definitions."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import re
+
+import astroid
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+from pylint.lint import PyLinter
+
+
+@dataclass
+class TypeHintMatch:
+    """Class for pattern matching."""
+
+    module_filter: re.Pattern
+    function_name: str
+    arg_types: dict[int, TypeChecker]
+    return_type: TypeChecker
+
+
+@dataclass
+class TypeChecker:
+    """Class for pattern matching."""
+
+    expected_type: str | None
+
+    def check(self, node: astroid.NodeNG) -> bool:
+        """Check the argument node against the expected type."""
+        if self.expected_type is None:
+            if isinstance(node, astroid.Const) and node.value is None:
+                return True
+            return False
+        if isinstance(node, astroid.Name) and node.name == self.expected_type:
+            # Name occurs when a namespace is not used, eg. "HomeAssistant"
+            return True
+        if isinstance(node, astroid.Attribute) and node.attrname == self.expected_type:
+            # Attribute occurs when a namespace is used, eg. "core.HomeAssistant"
+            return True
+        return False
+
+
+_MODULE_FILTERS: dict[str, re.Pattern] = {
+    "init": re.compile(r"^homeassistant.components.[a-zA-Z_]+$"),
+}
+
+_ARGUMENT_MATCH: dict[str, TypeChecker] = {
+    "HomeAssistant": TypeChecker("HomeAssistant"),
+    "ConfigType": TypeChecker("ConfigType"),
+    "ConfigEntry": TypeChecker("ConfigEntry"),
+}
+_RETURN_MATCH: dict[str, TypeChecker] = {
+    "bool": TypeChecker("bool"),
+    "None": TypeChecker(None),
+}
```

Comment:
These mappings don't save anything except maybe few bytes of memory for the cost of increased complexity. And they are even longer to type:
```
TypeChecker("ConfigType")
vs
_ARGUMENT_MATCH["ConfigType"]
and
TypeChecker("bool")
vs
_RETURN_MATCH["bool"]
```
Just use `TypeChecker` in the first place.

Reply:
I create TypeChecker to map a hint to a message. How that the message are standard I think I can drop the TypeChecker class altogether